### DISCUSSION TITLE
Improve usability of `.returns` for extending response data type.

### DISCUSSION THREAD

**Discussion by emmnull** [supabase/supabase#12238] [Feature Requests]
While the `.returns<T>()` method may be considered as a temporary patch until better overall type inference for joined row(s) lands (https://github.com/supabase/postgrest-js/issues/303), its current form more often than not leads to cumbersome and annoying code repetition if one wants to augment or patch the initially inferred return type instead of simply replacing it. I think `.returns` (or a new reponse method) should provide a way to modify the expected return type without having developpers redeclare it manually.

For an example, here's a (far-from-perfect) helper I drafted and that provides something similar to the requested behavior:

```ts
type Single<T> = T extends readonly unknown[] ? T[0] : never;

/**
 * As it stands, typescript can't do partial inference of generics :(. We are thus currying through
 * buff() to provide a common type inference site.
 */
export function buff<D>(data: D) {
  type SingleD = Single<D>;
  function singularize<Buff extends Partial<{ [K in keyof SingleD]: BuffOptions }>>() {
    type TBuff = {
      [K in keyof Buff]: K extends keyof SingleD
        ? Buff[K] extends 'single'
          ? Single<SingleD[K]>
          : Buff[K] extends 'maybeSingle'
          ? Single<SingleD[K]> | null
          : Buff[K]
        : never;
    };
    type S = Omit<SingleD, keyof Buff> & TBuff;
    return data as D extends unknown[] ? S[] : S;
  }

  return { singularize };
}
```
Usable like so:
```ts
const data = db
    .from('table')
    .select(
      `
        *,
        type:project_type(*),
        banner:projects_images!banner_id(*),
        gallery:projects_images!project_id(*),
        publication_status:projects_publication_status(*)
      `
    )
    .order('updated_at', { ascending: false })
    .range(...pagination(0, 10))
    .then((res) => {
      if (res.error) {
        // handle error with framework-specific helpers
      }
      return buff(res.data).singularize<{ publication_status: 'single' }>();
    });
```

**Comment by soedirgo** [comment]
Thanks for posting this, this seems like a valid concern to me. We could use something like [this](https://github.com/sindresorhus/type-fest/blob/main/source/merge-deep.d.ts) to make it work for nested foreign tables. We want to keep the default behavior of `.returns()` as-is though, to prevent breaking changes. How do you prefer the DX to look like?

```
await supabase
  .from('table')
  .select('*, foreign(*)')
  .returns<{ foreign: Foreign | null }[], { merge: true }>()
```

or

```
await supabase
  .from('table')
  .select('*, foreign(*)')
  .overrideTypes<{ foreign: Foreign | null }[]>()
```

or something else?

**Comment by emmnull** [comment]
@soedirgo, I would lean towards the second syntax (the dedicated method). But then again, If I understand correctly, this would still require to define the joined data's type from scratch? More often than not, the only _broken_ thing I encounter pertains to the expected cardinality for nested types, types that are otherwise correct (I generally override any problematic generated types - like erronous or missing column return types - globally, before passing them to `createClient`).

I'm still unsure what the better approach would be in terms of ergonomics and extendability. Maybe a string-parsing type function like my example is not so desirable (I know I myself don't really like magic strings like `single` or `maybeSingle`). Could the client maybe expose a collection of chainable type modifiers serving different purposes? (`returnType`, `extendType`, `singleType`, ...?)
